feat: User agreements API for generic agreement records

xitij2000 opened the pull request:
## Description

This change adds a new kind of generic user agreement that allows plugins or even the core platform to record a user's acknowledgement of an agreement.

## Supporting information

Product Review: https://github.com/openedx/platform-roadmap/issues/465

## Testing instructions

- This creates a new REST APIs for managing agreements:
- The API at /api/agreements/v1/agreement/ will list user agreements.  /api/agreements/v1/agreement/<type>/ will return details about an agreement of a particular type and  /api/agreements/v1/agreement/<type>/text will return the full licence / agreement text
-  The API at /api/agreements/v1/agreement_record/<agreement_type> will return the current user's agreement acceptance status. POSTing to it will cause that agreement type to be accepted. 


To test with the [corresponding frontend change](https://github.com/openedx/frontend-app-authoring/pull/2845) you need a few things:
- Create a user agreement using the django admin
- Use the MFE config API to configure gating based on acceptance as follows:
```json
{
  "AGREEMENT_GATING": {
    "uploads.video": ["abc_xyz"]
  }
}
```
This is essentially saying the to access the "uploads.video" feature you need to accept an agreement with the type "abc_xyz" so fill in the id from the agreement created before. 

Now the UI should show an alert to accept the agreement, and the video upload UI will be blurred out and blocked till the agreement is accepted. The other gating options are "upload.files" and "uploads". The idea being that if a user accepts an agreement that unblocks uploads it will automatically unblock files and videos but there is the option of being more granular.

## Deadline

"None" 


Private-ref: https://tasks.opencraft.com/browse/BB-10395

xitij2000 committed:
feat: User agreements API for generic agreement records

This change adds a new kind of generic user agreement that allows plugins or
even the core platform to record a user's acknowledgement of an agreement.

kaustavb12 commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +238,50 @@ def _user_signature_out_of_date(username, course_id):
+    Creates a user agreement record if one doesn't already exist, or updates existing
+    record to current timestamp.
+    """
+    record, _ = UserAgreementRecord.objects.update_or_create(
```
Since the primary use of this record would be for generating complaince reports, I feel we should only create records here and not update them. This would also be useful if, for example, some user agreement needs to be accepted periodically.

xitij2000 replied:
The idea here is that if you have an agreement, let's say "termsofservice", that can be updated, you can check if a user has accepted the agreement and also if they've accepted the latest agreement. For a report you can simply check which user's have a record for the agreement created after the date of the new agreement. If it needs to be accepted periodically then you can simply check in the frontend that the agreement was accepted in the current period.

kaustavb12 replied:
> The idea here is that if you have an agreement, let's say "termsofservice", that can be updated, you can check if a user has accepted the agreement and also if they've accepted the latest agreement

We can do that just the same even if we keep all records and not update them. No ?

In fact, specifically for the "terms of sevice" example, keeping all old records would enable someone doing an audit to determine under which version of the terms did an old user activity (for example if the user uploaded copyright protected files) take place.

I can see how old records can be retained even with `update_or_create` function by perhaps changing the agreement type for each version of the agreement, but it seems right to retain things like agreement records instead of updating them.

xitij2000 replied:
I don't think automatically changing the agreement type is a good idea. It's definitely an option that can be adopted by consumers of these APIs. For instance, you could have an agreement type of "toc-2024" and then "toc-2026" and they will both be tracked separately. 

However, I guess I'm just not seeing what you're talking about as a realistic scenario in this case. What you are talking about is a more complex requirement for auditing and I don't think this is trying to be that. 

What I'm looking at is a few simple scenarios. A ToS agreement, a fair use agreement etc. All you want to know is if the person has agreed to the current form of the agreement. You can do that by recording the date on which the updated agreement was posted and comparing against the date that the user last accepted the agreement. If they accepted after the latest agreement was posted, then they are good to go, otherwise, ask them to accept the agreement again. 

If you want to, you can make the logic more complex and check if there have been multiple updates to the agreement since their last acceptance, and if so, show then a UI that shows the differences between versions. Or at the very least show an "our terms have changed" UI.

In the scenario you shared, it probably doesn't matter in what duration those files were posted, as long as they were posted before the current agreement, because they might not align with it.

I don't think it would be a big deal or a lot of effort to adapt this code to keep one record for each acceptance, but currently this will scale with users * agreements and with your proposed change it will scale with users * agreements * agreement change and someone could simply code a loop posting agreements and flood the db. 

I guess they still can because the agreement type is not validated, so perhaps that is one way to go next.

kaustavb12 replied:
> However, I guess I'm just not seeing what you're talking about as a realistic scenario in this case. What you are talking about is a more complex requirement for auditing and I don't think this is trying to be that.

I was basing my assumption about auditing requirements based on this portion of the client's requirements which triggered this PR :
> - Log user interactions with the agreement (e.g., acceptance date) for auditing and compliance.
> - Enable report generation to check who has accepted the agreement within a given timeframe. 

It is a bit unclear in these statements as to how complex these "auditing and compliance" requirements are, and thought maintaining all records would be a safer option (without needing too much changes to the current implementation) which would provide users the flexibility to go beyond simple acceptance checking if need be.

But for now I guess I am ok with moving forward with the implementation as is. We can always change it later if so required.

xitij2000 replied:
I interpreted that as log the acceptance date not log every acceptance date, however, maybe I'm wrong and we can consult with them to clarify. 

For the report generation they only wanted that it be possible for them to do so using SQL on their side.

xitij2000 replied:
I've updated it to create a new record each time.

kaustavb12 replied:
Thank you !! LGTM :+1:

kaustavb12 commented on openedx/core/djangoapps/agreements/serializers.py:
```diff
@@ -31,3 +32,10 @@ class LTIPIISignatureSerializer(serializers.ModelSerializer):
+    """
+    Serializer for UserAgreementRecord model
+    """
+    accepted_at = serializers.DateTimeField()
```
Why not include user and agreement_type too ?
I realize the user and agreement type are infered from the request itself, but I feel that when we are doing the GET request, we are fetching a record of the user agreement and not just the timestamp(s) of when the user agreed. Including the username and agreement type would make the fetched record complete.

What do you think ?

xitij2000 replied:
Sure, that makes sense.

kaustavb12 commented:
:+1:

- [x] I tested this: Tested in local setup
- [x] I read through the code

xitij2000 commented:
@mphilbrick211 We were just discussing internally about what the process for reviewing such a PR should be. It doesn't fall into the product review category, but it does introduce a new API.

feanil commented:
@xitij2000 there is the API but I think the fact that we're introducing a model that tracks arbitrary agreements is a product feature that should get some sort of product review,  your open question about whether the content of the agreement should be stored here or not is a good one.  I could also imagine having the aggreements stored in their own model that this model foreign keys to for efficiency also.  The other aggerements being tracked here are explicity so I think discussing why it's useful for the platform to support generic aggrements in the core is an open question.

xitij2000 commented:
@feanil Sure, I will allocate some time to move this through product review. 

Re: Storing agreements in a model. For our client's use case, this would need to be a link to an agreement page, in which case having that being in siteconfiguration  was sufficient. Perhaps instead it can be an Agreement that can either have text or an external link with an update date. 

In any case I'll allocate some time next sprint for this.

feanil commented:
Another question to answer in your product review is why this needs to be a core product feature rather than a backend plugin that introduces the new agreements models you need?

xitij2000 commented:
@feanil  I don't think it particularly **needs** to be a core product feature other than that being ideal if such a feature is more widely desired. This would be pretty easy to extract out into a plugin.

feanil commented:
@xitij2000 I don't think we need this as a core product feature right now, let's make it a plugin for now and if we need to make updates to the core we can do that in order to enable this plugin.

itsjeyd commented:
@feanil Would there be an option for the plugin repo to be added to the openedx organization in the medium/long term? Speaking in terms of OEP-57, I guess another way to phrase this question would be, do you see a path forward where the plugin could become an official extension?

CC @xitij2000 @viadanna @kaustavb12 @cassiezamparini

feanil commented:
@itsjeyd yes, I could see a path if this was something that the Product team wanted. I think to make that decision this would have to go through the product proposal process.

itsjeyd commented:
Thanks @feanil, we'll go that route.

mphilbrick211 commented:
Hi @itsjeyd @xitij2000! Is this still in progress?

xitij2000 commented:
@mphilbrick211 We currently have a product proposal being worked on for this. I will post the link here once one exists.

feanil commented:
@xitij2000 I moved this to draft until the product proposal process concludes.

itsjeyd commented:
Product review for this PR is complete.

Product review ticket: https://github.com/openedx/platform-roadmap/issues/465

@xitij2000 Could you link to the product review ticket from the PR description as well? (And remove the Open Questions section while you're at it, it's probably not needed any longer?)

xitij2000 commented:
> Product review for this PR is complete.
> 
> Product review ticket: [openedx/platform-roadmap#465](https://github.com/openedx/platform-roadmap/issues/465)
> 
> @xitij2000 Could you link to the product review ticket from the PR description as well? (And remove the Open Questions section while you're at it, it's probably not needed any longer?)

I've made the updates, but I'm not clear on the next step here. Is this approved for merging? Or do we need to move it to a plugin?

itsjeyd commented:
@xitij2000 

> I've made the updates, 

Thanks!

> but I'm not clear on the next step here. Is this approved for merging? Or do we need to move it to a plugin?

Right now, neither. Please check [this update](https://tasks.opencraft.com/browse/BB-9330?focusedCommentId=319631&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-319631) and the conversations/comments that it links to.

xitij2000 committed:
feat: New User Agreements API

Adds new models and API to store user agreements such as fair use agreements,
terms of service, code of conduct etc. that need to be accepted by the user.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/serializers.py:
```diff
@@ -31,3 +32,24 @@ class LTIPIISignatureSerializer(serializers.ModelSerializer):
+    updated = serializers.DateTimeField(read_only=True)
+
+
+class UserAgreementRecordSerializer(serializers.Serializer):
```
@xitij2000  any specific reason for not using ModelSerializer? I think it might help here.

farhaanbukhsh replied:
@xitij2000 ^ did you miss this comment by mistake?

xitij2000 replied:
I guess so. I find this a lot more explicit in terms of the fields to include or exclude. You can do that with ModelSerializer and `read_only_fields` etc. but this makes things a lot more explicit. I don't have any big objection to making it a model serialiser though.

farhaanbukhsh replied:
I think let's use ModelSerializer, this will help us keep the codebase uniform.

xitij2000 replied:
I'm trying to switch over but the thing is the serializer is mostly being used with data classes not the model directly, so I need to rewrite bigger bits of the API than simply deriving from a different base and calling it done.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -2,21 +2,30 @@
+from .models import UserAgreement
+from .serializers import IntegritySignatureSerializer, LTIPIISignatureSerializer, UserAgreementRecordSerializer, \
+    UserAgreementSerializer
+from ...lib.api.view_utils import view_auth_classes
```
Why are we changing this to relative imports?

xitij2000 replied:
Not sure, other than maybe my ide is configured that way.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -159,3 +168,133 @@ def post(self, request, course_id):
+        if record is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+        serializer = UserAgreementRecordSerializer(record)
+        return Response(serializer.data)
```
```suggestion
        return Response(serializer.data, status=...)
```

we should pass status here.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -159,3 +168,133 @@ def post(self, request, course_id):
+        # if agreement is None:
+        #     return Response(status=status.HTTP_404_NOT_FOUND)
+        serializer = UserAgreementSerializer(agreements, many=True)
+        return Response(serializer.data)
```
Status missing here!

https://github.com/openedx/openedx-platform/pull/35895/changes#r2736307512

What do we do when there is no list?

farhaanbukhsh commented on openedx/core/djangoapps/agreements/serializers.py:
```diff
@@ -3,9 +3,10 @@
-from openedx.core.djangoapps.agreements.models import IntegritySignature, LTIPIISignature
 from openedx.core.lib.api.serializers import CourseKeyField
 
+from .models import IntegritySignature, LTIPIISignature
```
Relative imports?

farhaanbukhsh commented on openedx/core/djangoapps/agreements/models.py:
```diff
@@ -70,3 +71,57 @@ class ProctoringPIISignature(TimeStampedModel):
+class UserAgreement(models.Model):
+    """
+    This model stores agreements that as user can accept that can gate certain
+    platform features.
```
```suggestion
    This model stores agreements that the user can accept, which can gate certain
    platform features.
```

farhaanbukhsh commented on openedx/core/djangoapps/agreements/models.py:
```diff
@@ -70,3 +71,57 @@ class ProctoringPIISignature(TimeStampedModel):
+    type = models.CharField(max_length=255, unique=True)
+    name = models.CharField(
+        max_length=255,
+        help_text='Human-readable name for the agreement type. Will be displayed to users in alert to accept the agreement.',
```
```suggestion
        help_text='Human-readable name for the agreement type. Will be displayed to users in an alert to accept/reject the agreement.',
```

farhaanbukhsh commented on openedx/core/djangoapps/agreements/models.py:
```diff
@@ -70,3 +71,57 @@ class ProctoringPIISignature(TimeStampedModel):
+    )
+    summary = models.TextField(
+        max_length=1024,
+        help_text='Brief summary of the agreement content. Will be displayed to users in alert to accept the agreement.',
```
```suggestion
        help_text='Brief summary of the agreement content. Will be displayed to users in an alert to accept the agreement.',
```

farhaanbukhsh replied:
Is this a required field?

xitij2000 replied:
Yes, we need it to display an alert.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/data.py:
```diff
@@ -1,8 +1,13 @@
+
 import attr
 
+from .models import UserAgreementRecord, UserAgreement
```
Relative import!

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -3,17 +3,15 @@
-from .data import LTIToolsReceivingPIIData
-from .data import LTIPIISignatureData
+from .data import LTIPIISignatureData, LTIToolsReceivingPIIData, UserAgreementRecordData
+from .models import IntegritySignature, LTIPIISignature, LTIPIITool, UserAgreementRecord
```
We should avoid relative import and go with absolute imports

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +238,45 @@ def _user_signature_out_of_date(username, course_id):
+    """
+    Retrieves all the agreements that the specified user has acknowledged.
+    """
+    for agreement_record in UserAgreementRecord.objects.filter(user=user).select_related("agreement"):
```
```suggestion
    for agreement_record in UserAgreementRecord.objects.filter(user=user).select_related("agreement", "user"):
```

Since we access the username in the dataclass, if we don't use this, there will be multiple queries to fetch users.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/models.py:
```diff
@@ -70,3 +71,57 @@ class ProctoringPIISignature(TimeStampedModel):
         app_label = 'agreements'
+
+
+class UserAgreement(models.Model):
```
Should we use "TimeStampedModel" here, which will give us created and updated dates

xitij2000 replied:
Not really, we want the creation timestamp to be automatic but not the update timestamp. The update timestamp will invalidate old agreements, i.e. if a user accepts an agreement, then updating the agreement will invalidate the old acceptance. 
Someone might want to make small formatting or spelling fixes without asking users to accept again. So they will need to set the update time manually. 

I also considered adding an extra form field which allows overriding whether this is a update that will need re acceptance. In either case the update date should be be applied automatically.

xitij2000 replied:
Perhaps there should be a created, updated, major update timestap or similar.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +238,45 @@ def _user_signature_out_of_date(username, course_id):
         return user_lti_pii_signature_hash != course_lti_pii_tools_hash
+
+
+def get_user_agreement_records(user: User) -> Iterable[UserAgreementRecordData]:
```
```suggestion
def get_user_agreement_records(user: User) ->Iterator[UserAgreementRecordData]:
```

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +238,45 @@ def _user_signature_out_of_date(username, course_id):
+        record = record_query.latest("timestamp")
+        return UserAgreementRecordData.from_model(record)
+    except UserAgreementRecord.DoesNotExist:
+        return None
```
```suggestion
        user_agreement_records = None
        record_query = UserAgreementRecord.objects.filter(
            user=user,
            agreement__type=agreement_type,
        )
        record = record_query.latest("timestamp")
        if record.exists():
           user_agreement_records = UserAgreementRecordData.from_model(record)
        return user_agreement_records
```

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +238,45 @@ def _user_signature_out_of_date(username, course_id):
+def create_user_agreement_record(user: User, agreement_type: str) -> UserAgreementRecordData:
+    """
+    Creates a user agreement record if one doesn't already exist, or updates existing
+    record to current timestamp.
```
If we are creating or updating, should we use create_or_update instead? ref: https://docs.djangoproject.com/en/6.0/ref/models/querysets/#update-or-create

xitij2000 replied:
I think the function documentation is incorrect. We never remove old timestamps since they maintain historical acceptance. Instead, a new record is created each time.

farhaanbukhsh reviewed (changes requested):
I have left few comments possibly we can squash the migrations in one.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/data.py:
```diff
@@ -21,3 +26,45 @@ class LTIPIISignatureData:
     lti_tools_hash: str
+
+
+@dataclass
```
@xitij2000 I think it will be consistent to use attrs instead of dataclasses.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -159,3 +172,134 @@ def post(self, request, course_id):
+            404: "Not Found",
+        },
+    )
+    def retrieve(self, request, pk=None):
```
```suggestion
    def retrieve(self, request, type=None):
```

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -159,3 +172,134 @@ def post(self, request, course_id):
+        """
+        Get the user agreement for this agreement type.
+        """
+        agreement = UserAgreement.objects.filter(type=pk).get()
```
```suggestion
        agreement = UserAgreement.objects.filter(type=type).get()
```

farhaanbukhsh reviewed (changes requested):
I have added the comment and requested a change, which looks like a bug to me.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/views.py:
```diff
@@ -159,3 +171,142 @@ def post(self, request, course_id):
+        """
+        Get the user agreement for this agreement type.
+        """
+        print(f"{kwargs=}")
```
@xitij2000 print statement we need to remove

farhaanbukhsh commented on openedx/core/djangoapps/agreements/toggles.py:
```diff
@@ -3,6 +3,7 @@
 """
 
 from opaque_keys.edx.keys import CourseKey
+
```
if we are not touching this file then let's not make any changes here.

farhaanbukhsh replied:
@xitij2000 ^ ping

xitij2000 replied:
Done.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/urls.py:
```diff
@@ -14,4 +23,6 @@
         course_id=settings.COURSE_ID_PATTERN
     ), LTIPIISignatureView.as_view(), name='lti_pii_signature'),
+    path("agreement_record/<slug:agreement_type>", UserAgreementRecordsView.as_view(), name="user_agreement_record"),
+    path("", include(router.urls)),
```
in the above line, we have used single quotes. Let's keep that as is and use single quotes.

xitij2000 replied:
AFAIK black has been adopted for some openedx repos, which consistently uses double quotes, which is why I favour them in new code to avoid future changes during reformatting.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/models.py:
```diff
@@ -30,12 +32,13 @@ class LTIPIITool(TimeStampedModel):
 
     class Meta:
-        app_label = 'agreements'
+        app_label = "agreements"
```
single quotes have been used all over the codebase let us keep it and not introduce variations here.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/api.py:
```diff
@@ -240,3 +248,46 @@ def _user_signature_out_of_date(username, course_id):
+def get_latest_user_agreement_record(
+    user: User,
+    agreement_type: str,
+) -> Optional[UserAgreementRecordData]:
```
This will not be optional right? What data flow will lead us to return None?

farhaanbukhsh commented on openedx/core/djangoapps/agreements/tests/test_views.py:
```diff
@@ -289,3 +291,68 @@ def test_post_lti_pii_signature(self):
+    def login(self):
+        self.client.login(username="testuser", password="password")
+
+    def test_get_user_agreement_record_for_missing_agreement(self):
```
We should be adding docstring to all these tests.

farhaanbukhsh commented on openedx/core/djangoapps/agreements/tests/test_api.py:
```diff
@@ -186,3 +190,52 @@ def _assert_ltitools(self, lti_list):
+            updated=datetime.now(),
+        )
+
+    def test_get_user_agreements(
```
I think we should add docstrings here.

farhaanbukhsh reviewed (changes requested):
I have some nits @xitij2000 and thanks a lot for the persistence, I think we are getting close to complete it.

farhaanbukhsh reviewed (approved):
👍

Just one more nit and also should be include the user agreement record in admin as well? Thank you for the persistent work on the PR

✅ I tested this
- By using the new frontend in authoring
- created a new agreement
- Added the agreement using MFE Config API
- Checking by creating multiple users
✅ I read through the code
❌  I checked for accessibility issues
❌  Includes documentation
